### PR TITLE
mariadb: support multiple versions

### PR DESCRIPTION
**Pull request by ajs124**
###### Motivation for this change
As with PostgreSQL, not every release of software is always compatible with the newest release of MariaDB.

Or maybe you just like an older version of MariaDB better for some other reason.
In any case, we have historically only been offering one version of MariaDB at a time, mainly because nobody implemented it, AFAICT.

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -21216,10 +21216,15 @@ with pkgs;
     asio = asio_1_10;
   };
 
-  mariadb = callPackage ../servers/sql/mariadb {
+  inherit (callPackage ../servers/sql/mariadb {
     inherit (darwin) cctools;
     inherit (darwin.apple_sdk.frameworks) CoreServices;
-  };
+  })
+    mariadb_104
```
Comment (aanderse) [inline pkgs/top-level/all-packages.nix:21223]: I've noticed more of a pattern in nixpkgs that would have us choose `mariadb_10_4`... but maybe I'm off base?

**Review by aanderse — APPROVED** [review]
This is awesome! Thanks for putting in the effort here.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (ajs124) [reply pkgs/top-level/all-packages.nix:21223]: No idea which way we normally do this, tbh. Do versionOlder and versionAtLeast understand versions with _ in them?

File: pkgs/servers/sql/mariadb/default.nix [hunk 2]
```diff
@@ -14,36 +14,37 @@
...
-    curl libiconv ncurses openssl pcre2 zlib
-  ] ++ optionals stdenv.hostPlatform.isLinux [ libkrb5 liburing systemd ]
-    ++ optionals stdenv.hostPlatform.isDarwin [ CoreServices cctools perl ]
-    ++ optional (!stdenv.hostPlatform.isDarwin) [ jemalloc ];
+    curl libiconv ncurses openssl zlib
+  ] ++ (packageSettings.extraBuildInputs or [])
+    ++ lib.optionals stdenv.hostPlatform.isLinux [ libkrb5 systemd ]
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [ CoreServices cctools perl ]
+    ++ lib.optional (!stdenv.hostPlatform.isDarwin) [ jemalloc ]
+    ++ (if (lib.versionOlder version "10.6") then [ libaio ] else [ liburing ])
```
Comment (ajs124) [inline pkgs/servers/sql/mariadb/default.nix:46]: uring is only available on Linux. I forgot to keep the check for that here. Let's not break Darwin again.

**Comment by ajs124** [comment]
Huh, and apparently just in time: https://mariadb.com/de/resources/blog/delivering-faster-innovation-to-mariadbs-community/

File: pkgs/servers/sql/mariadb/default.nix [hunk 3]
```diff
@@ -188,38 +195,54 @@ server = stdenv.mkDerivation (common // {
...
     rm -r "$out"/OFF
   '';
 
-  CXXFLAGS = optionalString stdenv.hostPlatform.isi686 "-fpermissive";
+  CXXFLAGS = lib.optionalString stdenv.hostPlatform.isi686 "-fpermissive";
 });
-in mariadb
+in {
+  mariadb_104 = mariadbPackage {
+    # Supported until 18 June 2024
```
Comment (flokli) [inline pkgs/servers/sql/mariadb/default.nix:233]: Use June 18, 2024, or just 2024-06-18 to make things simple.

File: pkgs/servers/sql/mariadb/default.nix [hunk 4]
```diff
@@ -188,38 +195,54 @@ server = stdenv.mkDerivation (common // {
...
+    version = "10.4.22";
+    sha256 = "sha256:000ca1hdnj2jg051cjgdd2ralgwgh2p8nwb1x6b85202xdpc7ga4";
```
Comment (flokli) [inline pkgs/servers/sql/mariadb/default.nix:235]: what's that hash format?

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (aanderse) [reply pkgs/top-level/all-packages.nix:21223]: OK. As is should work fine then :+1:

**Review by flokli — COMMENTED** [review]
The diff is pretty horrible to look at - I couldn't really review all of it.

This definitely needs an addition to the upcoming release notes, so people know and can decide which mariadb version to pick.

**Comment by Izorkin** [comment]
@ajs124 small cleanup:
```
diff --git a/pkgs/servers/sql/mariadb/default.nix b/pkgs/servers/sql/mariadb/default.nix
index 1d8a64b5851..3f1f3448dad 100644
--- a/pkgs/servers/sql/mariadb/default.nix
+++ b/pkgs/servers/sql/mariadb/default.nix
@@ -2,7 +2,7 @@
 # Native buildInputs components
 , bison, boost, cmake, fixDarwinDylibNames, flex, makeWrapper, pkg-config
 # Common components
-, curl, libiconv, ncurses, openssl, pcre2, pcre
+, curl, libiconv, ncurses, openssl, pcre, pcre2
 , libkrb5, libaio, liburing, systemd
 , CoreServices, cctools, perl
 , jemalloc, less
@@ -41,7 +41,7 @@ commonOptions = packageSettings: rec { # attributes common to both builds
     curl libiconv ncurses openssl zlib
   ] ++ (packageSettings.extraBuildInputs or [])
     ++ lib.optionals stdenv.hostPlatform.isLinux ([ libkrb5 systemd ]
-      ++ (if (lib.versionOlder version "10.6") then [ libaio ] else [ liburing ]))
+    ++ (if (lib.versionOlder version "10.6") then [ libaio ] else [ liburing ]))
     ++ lib.optionals stdenv.hostPlatform.isDarwin [ CoreServices cctools perl ]
     ++ lib.optional (!stdenv.hostPlatform.isDarwin) [ jemalloc ]
     ++ (if (lib.versionOlder version "10.5") then [ pcre ] else [ pcre2 ]);

```

**Comment by Izorkin** [comment]
Neded adding mariadb 10.3? Support to May 2023.

**Comment by ajs124** [comment]
I tried that, but ran into issues. I mainly wanted this so we can do it going forward. If anyone actually needs 10.3, I'd try and add it again, but if nobody does, then I think we can just omit it.

File: nixos/tests/nextcloud/default.nix [hunk 5]
```diff
@@ -1,21 +1,27 @@
...
+    "with-mysql-and-memcached-nc${toString ver.nextcloudVersion}-mariadb${toString ver.mariadbVersion}" = import ./with-mysql-and-memcached.nix {
       inherit system pkgs;
-      nextcloudVersion = ver;
+      inherit (ver) nextcloudVersion mariadbVersion;
     };
-  })
-{ }
-  [ 21 22 23 ]
+  }) { }
+(lib.cartesianProductOfSets { nextcloudVersion = nextcloudVersions; mariadbVersion = mariadbVersions; })
```
Comment (Ma27) [inline nixos/tests/nextcloud/default.nix:27]: I'm not sure if we're overcomplicating things here. Yes, it was a mistake from - well - my side to miss the MariaDB incompatibilities with certain Nextcloud versions, but in the end upstream's job to recommend compatible versions and we should follow that as much as possible. So what I'm suggesting is:

* Revert the change here
* Declare a working mariadb package depending on the Nextcloud version (i.e. Nextcloud 23 and later support 10.6[1], older majors use 10.5) and we only check if this combination from services.nextcloud works fine here.

[1] This is just what I think I recall, I haven't checked that, so no guarantee that this example is 100% correct :)

**Comment by Ma27** [comment]
LGTM except for the comment above.

Approve & merge should be done by someone who actually uses (or even maintains) mariadb here :)

File: nixos/tests/nextcloud/default.nix (hunk 5 above)
Comment (ajs124) [reply nixos/tests/nextcloud/default.nix:27]: Ok, so how do we suggest we proceed? I can just rebase out the nextcloud change and clean the history up while I'm at it. I'm not overly familiar with nextcloud module and which version of nextcloud is compatble with which mariadb version, so can you maybe do that change?
On the other hand, I'm not sure if we want to default the version of mariadb used on the whole system to something, based on the nextcloud version that's configured. mariadb also doesn't support clean downgrades AFAIR. upgrades also sort-of require manual intervention (running `mysql_upgrade`) rn. not sure if that's documented anywhere.

Comment (Ma27) [reply nixos/tests/nextcloud/default.nix:27]: How about setting it with `mkDefault` and perhaps even throw an eval warning if the default is used? (i.e. `mkDefault (lib.warn "using mariadb x.y for nextcloud" pkgs.mariadb_x_y)`?

Comment (ajs124) [reply nixos/tests/nextcloud/default.nix]: Hm. What would we `mkDefault` to though? 10.5 for all relases < 23? I'm not sure I can commit to maintaining that mapping, tbh. I've rebased the nextcloud changes out, for now.

**Comment by Izorkin** [comment]
If enable build mariadb 10.4 with storage tokudb needed jamalloc 4.5.0.

**Comment by ajs124** [comment]
> If enable build mariadb 10.4 with storage tokudb needed jamalloc 4.5.0.

We don't have that in nixpkgs anymore. What do you suggest? Disabling tokudb for older releases?

**Comment by dasJ** [comment]
We can also just drop 10.4, I don't think somebody has an issue with us not re-introducing an old version we didn't ship previously

**Comment by Izorkin** [comment]
TokuDB is already disabled by default:
```
"-DWITHOUT_TOKUDB=1"
```
If TokuDB is not needed by anyone, then you can leave it unchanged.

**Review by dasJ — APPROVED** [review]

**Comment by aanderse** [comment]
🎉 

Thanks everyone!

**Comment by dasJ** [comment]
Where is that used? I cannot find any uses in `nixos/` and `pkgs/`

**Comment by immae** [comment]
Nixos might not package this kind of applications, but it’s a common occurence in symfony (doctrine) websites for instance: https://symfony.com/doc/current/reference/configuration/doctrine.html#doctrine-dbal-configuration . It’s the equivalent of `postgresql.psqlSchema` for postgresql (although in mariadb it’s more a protocol information than a schema information)

**Pull request merged**